Source/Extensions/Blazorise.DataGrid/DataGrid.razor.cs

```diff
@@ -1788,12 +1788,85 @@ internal IEnumerable<DataGridColumn<TItem>> DisplayableColumns
+                .ToList();
+
+            var newOrderedDisplayColumns = new List<DataGridColumn<TItem>>();
+
+            for ( int i = 0; i < orderedDisplayColumns.Count; i++ )
+            {
+                var displayColumn = orderedDisplayColumns[i];
+                newOrderedDisplayColumns.Add( displayColumn );
+
+                if ( !string.IsNullOrWhiteSpace( displayColumn.HeaderGroupCaption ) && orderedDisplayColumns.Count > i + 1 )
+                {
+                    var toRemove = new List<DataGridColumn<TItem>>();
+                    foreach ( var remainingDisplayColumn in orderedDisplayColumns.Skip( i + 1 ) )
+                    {
+                        if ( remainingDisplayColumn.HeaderGroupCaption == displayColumn.HeaderGroupCaption )
+                        {
+                            newOrderedDisplayColumns.Add( remainingDisplayColumn );
+                            toRemove.Add( remainingDisplayColumn );
+                        }
+                    }
+                    orderedDisplayColumns.RemoveAll( x => toRemove.Contains( x ) );
+                }
+            }
+
+            return newOrderedDisplayColumns;
+
         }
     }
 
+    /// <summary>
+    /// Gets only columns that are available for display in the grid group header.
+    /// </summary>
+    internal IEnumerable<(DataGridColumn<TItem> col, int colSpan)> DisplayableHeaderGroupColumns
+    {
+        get
+        {
+            var orderedDisplayColumns = Columns
+                .Where( x => x.IsDisplayable || x.Displayable )
+                .OrderBy( x => x.DisplayOrder )
+                .ToList();
+
+            var newOrderedDisplayColumns = new List<(DataGridColumn<TItem> col, int colSpan)>();
+
+            for ( int i = 0; i < orderedDisplayColumns.Count; i++ )
+            {
+                var displayColumn = orderedDisplayColumns[i];
+                var colSpan = 1;
+
+                if ( !string.IsNullOrWhiteSpace( displayColumn.HeaderGroupCaption ) && orderedDisplayColumns.Count > i + 1 )
+                {
+                    var toRemove = new List<DataGridColumn<TItem>>();
+                    foreach ( var remainingDisplayColumn in orderedDisplayColumns.Skip( i + 1 ) )
+                    {
+                        if ( remainingDisplayColumn.HeaderGroupCaption == displayColumn.HeaderGroupCaption )
+                        {
+                            colSpan++;
+                            toRemove.Add( remainingDisplayColumn );
+                        }
+                    }
+                    orderedDisplayColumns.RemoveAll( x => toRemove.Contains( x ) );
+                }
+
+                newOrderedDisplayColumns.Add( (displayColumn, colSpan) );
+            }
+
+            return newOrderedDisplayColumns;
+
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets whether user can see group header column captions.
+    /// </summary>
+    internal bool IsGroupHeaderCaptionsEnabled

```

Comment:
What if we have `ShowHeaderGroupCaptions`, just as we have `ShowCaptions`? I know the user would have to enable it strictly, but then we have less work to check when rendering.

Reply:
I guess it's okay...